Core/include/Acts/TrackFinder/CombinatorialKalmanFilter.hpp

```diff
@@ -0,0 +1,1255 @@
+            }
+            trackStateProxy.jacobian() = jacobian;
+            trackStateProxy.pathLength() = pathLength;
+
+            // Get and set the type flags
+            auto& typeFlags = trackStateProxy.typeFlags();
+            typeFlags.set(TrackStateFlag::MaterialFlag);
+            typeFlags.set(TrackStateFlag::ParameterFlag);
+            if (isOutlier) {
+              // Set the outlier type flag
+              typeFlags.set(TrackStateFlag::OutlierFlag);
+            } else {
+              // Set the measurement type flag
+              typeFlags.set(TrackStateFlag::MeasurementFlag);
+            }
+
+            // Assign the source link and calibrated measurement to the track
+            // state
+            if (sourcelinkShared) {
+              // The source link is already stored, just set the index
+              auto sharedMeasurement =
+                  result.fittedStates.getTrackState(index_it->second);
+              trackStateProxy.data().iuncalibrated =
+                  sharedMeasurement.data().iuncalibrated;
+            } else {
+              trackStateProxy.uncalibrated() = sourcelinks.at(index);
+            }
+            std::visit(
+                [&](const auto& calibrated) {
+                  trackStateProxy.setCalibrated(calibrated);
+                },
+                m_calibrator(trackStateProxy.uncalibrated(),
+                             trackStateProxy.predicted()));
+
+            // Set the filtered parameter
+            if (isOutlier) {
+              // No Kalman update for outlier
+              // Set the filtered parameter index to be the same with predicted
+              // parameter
+              trackStateProxy.data().ifiltered =
+                  trackStateProxy.data().ipredicted;
+
+              ACTS_VERBOSE(
+                  "Creating outlier track state with tip = " << trackTip);
+
+              // Count the number of processedStates, measurements, outliers and
+              // holes
+              auto tipState =
+                  TipState{preTipState.nStates + 1, preTipState.nMeasurements,
+                           preTipState.nOutliers + 1, preTipState.nHoles};
+
+              // Check if need to stop this branch
+              if (not m_branchStopper(result.fittedStates, trackTip)) {
+                // Remember the active tip and its state
+                result.activeTips.emplace(trackTip, std::move(tipState));
+                // Count the valid branches on current surface
+                nBranchesOnSurface++;
+              }
+            } else {
+              // If the update is successful, update the tip state and count the
+              // states on surface
+              auto updateRes = m_updater(state.geoContext, trackStateProxy);
+              if (!updateRes.ok()) {
+                ACTS_ERROR("Update step failed: " << updateRes.error());
+                return updateRes.error();
+              } else {
```

Comment:
You probably don't need this `else` (also the case in some other spots), since you return above if there is an error, it should be safe to use the result and assume ok. This could get rid of some levels of indentation, and might improve readability.

Reply:
Exactly!